Implement PHP DI container and refactor.  Also implements new Asset data interface for extendable settings passed to js.

nerrad opened the pull request:
This work addresses an alternative to #936 and https://github.com/woocommerce/woocommerce-admin/pull/2867.

It is a companion to https://github.com/woocommerce/woocommerce-admin/pull/2917

As I've reviewed the current plugin code there a few of things that stood out to me:

- server provided settings (not via api, embedded on the loaded view via a js global) are shared with the woocommerce-admin repository which is not yet in WC core, whereas blocks is.  The current implementation is a bit fragile and requires loading the global on every route.  There is need for an interface for not only exposing these settings but also adding to them. This will become increasingly important as we start doing more rendering of blocks on the frontend and have need of hydrating the initial view via data provided from the server via this settings object.
- Blocks registration and initialization has a lot of manual pieces (configuration array etc) that introduce greater potential for human error.  
- The plugin is primarily using static classes for various initialization pieces.  This works but is not great structure for purposes of testing and dependencies between classes (not to mention some cases of multisite use).  Typically this is used for easier global access to the class functionality (they are essentially namespaced global functions).  How they are used here is for the most part okay because they are stateless, but I think we'll still want to implement a system for classes that will have state (i.e. the `AssetDataRegistry` I introduced in this pull) and for managing shared instances.

With that in mind, this pull does the following:

### Introduction of a simple Dependency Injection Container.

While I could have simply pulled from existing options out there, I don't think we need a full blown complicated DIC right now.  If this pattern proves useful - it may be something we'd want to implement in WC core at some point.  I'm assuming the reader knows what a DIC is, if not ask in the pull comments and I'll link to some resources.

For now, I implemented a very basic DIC for use.  Some notes:

- Generally you don't reference the DIC in any lower level classes in the callstack and it is generally only exposed for use in factory/controllers/routers.  However, because there will be a transitionary process, the current DIC (`src/Registry/Container`) instance is exposed via `wc_blocks_container`.
- The interface is fairly similar to that of [Pimple](https://pimple.symfony.com/).  I didn't just use an existing solution because then we'd have to worry about conflicts with other plugins that might use the same solution.

#### Usage.

The DIC has the following basic interface:

**Registration** (`wc_blocks_container()->register( $identifier, $instantiation_callback )`)

You register a class with the DIC via the `register` method.  It receives two arguments: an identifier (usually you'd use the fully qualified class name), and a callback used for instantiating the class instance when it's requested.  The callback will receive the container as an argument so it can pull dependencies for the instantiation of its class.

```php
class SomeClassA {}

// register
$container = wc_blocks_container();
$container->register( SomeClassA::class, function () { return new SomeClassA; } );
```

**Retrieving** (`wc_blocks_container()->get( $identifier )` ).

The `get` method on the container will retrieve the instance of the class in the container for the given identifier.  There's a couple important points to remember here.

- a Class registered in the container is not instantiated until the first time the `get` is called for that registered class.
- by _default_ registered classes will only hold a single instance of a class.  This means subsequent calls to `get` for that class will return the same instance of the class.

With that last point, you may be wondering "How do I return a new instance every time I request the class?".  The way to do that is to use wrap your class instantiation callback in a call to `$container->factory()`.  So you'd end up with something like this:

```php
class SomeClassA {}

$container = wc_blocks_container();

$container->register(
    SomeClassA::class,
    $container->factory(
        function() { return new SomeClassA; }
    )
);

// the two below instances are not the same instance.
$instanceA = $container->get( SomeClassA::class );
$instanceB = $container->get( SomeClassA::class );
```

Putting things together, here's an example of registering a class that is dependent on another class.

```php
class SomeClassA {
    public function __construct( SomeClassB $b ) {}
}

class SomeClassB {}

$container = wc_blocks_container();
$container->register( 
    SomeClassA::class,
    function ( $passed_container ) {
        return new SomeClassA( $passed_container->get( SomeClassB::class ) );
    }
);
$container->register(
    SomeClassB:class,
    function() { return new SomeClassB(); }
);
```

### Introduction of `AssetDataRegistry`

The implementation of this new class in this pull is for a couple reasons:

- it provides an interface for registering data throughout the system by not only wc blocks, but also woocommerce-admin and potentially woocommerce core in the future.
- it provides an initial implementation example of using the DIC (which in turn required a minimal refactor of the bootstrap process).

The `AssetDataRegistry` is automatically registered in the DIC during bootstrap so it's available from the container.

**The interface for registering data to output on the `wcSettings` global for javascript**

> **Note:** Whatever data is registered via the interface during a request will be output on the current view if the `wc-settings` handle is a dependency for any enqueued script or enqueued directly itself.

Whenever you need to register data for output to the wcSettings global, you can use the following interface:

`Automattic\WooCommerce\Blocks\Assets\AssetDataRegistry::add( $key, $data )`

The method receives two arguments:

- `$key`: a unique identifier for the data being passed through.  This will be exposed on `wcSettings[ $key ]`.
- `$data`: the data you want attached to the global.  **Note:** if you want the data to be lazy loaded at the time of preparing the data for output (as opposed to generating on registration), you can provide a closure and the closure will be invoked when the data is needed on the current view.  This is really useful for data retrieved from the database for instance, and you want the latest data before outputting to the page.

#### Usage:

```php
// simple example of adding a plain value to the queued data.
// note, ideally you would not use `wc_blocks_container()` for retrieving the AssetDataRegistry and 
// instead would make it a dependency of whatever class is using it.  But for the purpose of keeping 
// this example simple I'm using it here.
$data_registry = wc_blocks_container()->get(
    Automattic\WooCommerce\Blocks\Assets\AssetDataRegistry::class
);
$data_registry->add( 'foo', 'bar' );

// example of delayed value generation
$data_registry->add( 'some_option', function() { return get_option( 'some_option' ); } );


// in javascript this would be available on `wcSettings.foo` and `wcSettings.some_option`
```

For back compat reasons and to make transitioning a bit easier, there is a filter that can be used for late adding of data in a request, `woocommerce_shared_settings` (this is currently being used in the [companion pull request in woocommerce-admin](https://github.com/woocommerce/woocommerce-admin/pull/2917)).  However, I'd like to discourage use of this filter because filters are notorious for easy breakage by add-ons/extensions so encouraging usage of the explicit interface helps to protect the integrity of the data passed through to the server.  Once our refactoring is done, I'd like to actually remove this filter.

### Javascript changes (introduction of `wc-settings` bundle)

The javascript side of the data refactor involved:

- Exporting to a bundle:
    - declared constants for more globally shared woocommerce data (`COUNTRIES`, `CURRENCY`, `ADMIN_URL` etc).
    - `getSetting` for interfacing with more dynamic data that is exposed via `wcSettings` api php side.
    - `setSetting` for js only setting of dynamic data used between modules (including client-side only mutation of setting state).
- aliasing `@woocommerce/settings` to this new interface (exposed on `wc.wcSettings`).
- aliasing `@woocommerce/block-settings` to a specific path which in turn is dependent on wcSettings (more block specific settings exposed via the AssetDataRegistry api).

The bundle itself exposes the new interfaces on `wc.wcSettings` global which can be accessed by other plugins (such as WC-A) via adding the `wc-settings` as a handle to the dependency array for any scripts wanting to use the interface.

**`getSetting( name, fallback, validation )`**

`getSetting` is an interface used to retrieve values from the setting state initialized by what was passed from the server via `wcSettings` global (via the `AssetDataRegistry`).  It has three arguments:

- `name`:  This will correspond to what setting you want to retrieve.
- `fallback`: This is a value that should be used if there is no value in the setting state for `$name`.
- `validation`: An _optional_ callback that can be used to validate/sanitize the setting value returned.  It will receive the setting value attached to `$name` in the setting state, and it is expected you would return the modified/unmodified value from the callback.

**`setSetting( name, value, validation )`**

`setSetting` is an interface used to _set_ values into the setting state for the given `name`.  **Note:** this _will_ override existing values for the given `name` but not persist them to the server.  So this should be used with care for existing values because that will affect any existing modules in the view relying on that setting.  In future iterations we may want to implement a mechanism for _protecting_ certain settings from being mutated.  It has three arguments:

- `name`: This will correspond to the setting you want to change in the setting state.
- `value`: The value to add to the state for the given name.
- `validation`: An optional callback for validating or sanitizing the setting that is added to the state.

### Back compat.

In this pull I've also taken care to cover back-compatibility scenarios:

- Via `BackCompatAssetDataRegistry` (which incidentally also demonstrates how you can do on the fly class substitution via the DIC), this will support earlier versions of `woocommerce/admin` without https://github.com/woocommerce/woocommerce-admin/pull/2917 in it.
- Also via `BackCompatAssetDataRegistry` the `woocommerce_components_settings` is supported (however this back compat is limited as it only factors in on earlier versions of WC_Admin or WC core.  Otherwise the filter is deprecated (hopefully there aren't too many plugins/extensions using the filter, however it is something that will need documented.

## Testing

* [ ] Smoke test all the blocks to make sure they still work as expected.
* [ ] Try this on a version of woocommerce/admin without https://github.com/woocommerce/woocommerce-admin/pull/2917 merged in.  Smoke test the functionality of WC admin and ensure there are no console errors, and things work as expected.
* [ ] Try this on a version of woocommerce/admin _with_ https://github.com/woocommerce/woocommerce-admin/pull/2917 merged in (or on that branch). Smoke test the functionality of WC admin and ensure there are no console errors, and things work as expected.

There are a number of phpunit tests that cover some of the php side changes.

## Next steps

* [ ] Get https://github.com/woocommerce/woocommerce-admin/pull/2917 merged.
* [ ] Continue refactoring, in smaller pulls, to utilize the DIC more broadly (and how blocks are loaded/initialized) (see #972 and #973)
* [ ] Plan out how this will be documented and shared on next merge to wc core.
* [ ] More broadly publicise the DIC and how to use (i.e. P2s), woo dev docs?

mikejolley commented on woocommerce-gutenberg-products-block.php:
```diff
@@ -14,6 +14,10 @@
  * @internal This file is only used when running the REST API as a feature plugin.
  */
 
+use Automattic\WooCommerce\Blocks\Domain\Bootstrap;
```
May be risky doing this before the PHP version check?

nerrad replied:
Nope. The only time this will cause a fatal is if run in a PHP 5.2 environment.  Chances of that are pretty slim considering it requires a perfect storm of also installing this on a site without WooCommerce core installed (if the user downloads the plugin separately) _and_ on a version of WordPress not supporting Gutenberg. [WooCommerce Admin also does this](https://github.com/woocommerce/woocommerce-admin/blob/9a64926d8af836493330236c2a1ce0e18ca064ce/woocommerce-admin.php#L22) and I figured if it was a problem, that would have changed by now (although from blame I see it's a recent change).

mikejolley replied:
Whilst it's true Woo core won't load this package < 5.3, as a feature plugin it could be installed and activated regardless. 

Thinking about it a different way, why bother with PHP version compare a few lines down if it's going to break before that code runs anyway? :) Ideally the check should run first to prevent a fatal.

nerrad replied:
> as a feature plugin it could be installed and activated regardless.

Very true, but how likely is it that people will install a plugin that has "blocks" in its name if they aren't running Gutenberg in their WordPress environment?

> Thinking about it a different way, why bother with PHP version compare a few lines down if it's going to break before that code runs anyway?

Because the version compare is for PHP5.6+ which covers things that are only available in PHP5.6+.  Namespaces and imports have been available since PHP5.3 so imo it's an edgecase.  With that said, I'll change things to use fully qualified class names instead if you want me to.  If we do so, we should probably be consistent and do the same in WC Admin as well.

mikejolley replied:
I prefer to err on the side of caution in the main file, so using fully qualified class names here would be my preference. Once we're sure PHP 5.6 is being used, throughout all other files, we're good to use whatever we need.

nerrad commented on assets/js/settings/shared/get-setting.js:
```diff
@@ -0,0 +1,29 @@
+export function getSetting(
+	name,
+	fallback = false,
+	validation = ( value ) => value
```
Hmm...

- Ya we should either provide `fallback` to the validation callback or run the callback on the fallback as well.  I think the former sounds better. Thoughts?
- I don't think it's necessary to split validation and sanitization callbacks.  The way I envisioned things working is that if something fails validation, then an error should be thrown (which means calling code would have to handle the error).  If we provide the fallback to the callback as a second arg, that also gives some more flexibility for how the consumer decides to handle failed validation (vs sanitization).  

One thing that could use improvement here is the name of the callback argument since it's not just intended for validation.  Any suggestions?

mikejolley replied:
- Is there an example of how an error would be raised from this, and caught?
- Not sure on naming. One benefit of supporting multiple callbacks is that you can be explicit in the naming i.e. santitize_callback, validate_callback, similar to how callbacks are added in the WP rest API. For a do-all callback, not really sure what you can call it...format? Just "callback"?
- Passing `fallback` is better IMO so you know when the fallback is used.

nerrad replied:
- renamed callback to `filter` and adjusted docs to remove references of sanitization/validation.  Essentially its a flexible callback for consumers to use how they wish for filtering the value (typically could be used for sanitization).
- pass thru `fallback` to the `filter` callback.  This allows for consumers to potentially do validation as well and return the fallback value if validation fails.

nerrad commented on src/Assets.php:
```diff
@@ -153,27 +96,28 @@ protected static function get_wc_block_data() {
+				'limitTags'          => $tag_count > 100,
+				'hasTags'            => $tag_count > 0,
+				'productCategories'  => $product_categories,
+				'homeUrl'            => esc_js( home_url( '/' ) ),
```
oh ya, `esc_url` would be better here 👍

nerrad commented on src/Assets/AssetDataRegistry.php:
```diff
@@ -0,0 +1,246 @@
+	public function add( $key, $data ) {
+		if ( ! is_string( $key ) ) {
+			if ( $this->debug() ) {
+				throw new InvalidArgumentException(
```
Ya I wondered about that.  However, this is _explicitly_ an error intended for developers - meaning, it should surface really quickly if there's a problem.  I'm wary that doing_it_wrong notices/errors/logging would result in inaction which could lead to unexpected behaviour if ignored.

mikejolley replied:
`_doing_it_wrong`/`wc_doing_it_wrong` are also for developers. They only surface visually if debug notices are turned on and displayed, otherwise it's logged. I don't think this is much different to what you have, besides it being a notice vs an exception. In both cases, debug is required otherwise it's ignored.

nerrad replied:
To clarify, when I wrote,

> I'm wary that doing_it_wrong notices/errors/logging would result in inaction which could lead to unexpected behaviour if ignored.

I mean that (from observation) due to how `doing_it_wrong` (wp core) has been used in various ways for everything from things that will "still work" but aren't correct implementation to things that wont' work - I decided that for here that it's important to actually break things (and throw an exception, because it's exceptional) - and fail catastrophically really early (assuming the developer is developing with debug mode on!).  It's less that the developer is "doing it wrong" and more that they have to choose a more unique key for the data property.

I think an improvement I can make here is to throw the exception always, but if in debug catch the exception and log it to `wc_caught_exception`.

nerrad replied:
I implemented a slight improvement in 504c48176.  However, in another iteration after this is merged, I'd like to explore having the try/catch at the bootstrap level and it can handle any bubbled exceptions appropriately.  For instance it could handle when request type is api or in debug mode and log and prepare for user-friendly message (if necessary).

nerrad replied:
Created https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/982 for this.

nerrad commented on src/Assets/BackCompatAssetDataRegistry.php:
```diff
@@ -0,0 +1,79 @@
+				'currency'      => array_merge(
+					$core_data['currency'],
+					[
+						'position'           => $core_data['currency']['symbolPosition'],
```
Yup saw your comment in the woo-admin pull and I like the idea.  We could have this happen on the js side where it automatically sets the prefix/suffix via the value of this property and exposes that for usage in the modules.

However, it sounds like something we could do in it's own pull?

nerrad replied:
Followup in https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/983

nerrad commented on src/Domain/Bootstrap.php:
```diff
@@ -0,0 +1,151 @@
+
+namespace Automattic\WooCommerce\Blocks\Domain;
+
+use Automattic\WooCommerce\Blocks\Assets as OldAssets;
```
Ya I already have an issue for tracking the removal: https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/973

Since I plan to do that fairly soon after this pull gets merged I'm not sure we need to do the name change right now (Ideally I'd like to get the refactor complete before next plugin release).

Regarding using `@todo`.  It appears our precommit hook flags those and prevents commits.  Is that something intended, or should we do a pull to tweak that?

mikejolley replied:
cc @Aljullu 

I think we should allow it. I got moaned at for using `TODO:` syntax before because of that rule :p

Aljullu replied:
> I think we should allow it. I got moaned at for using TODO: syntax before because of that rule :p

It sounds good to me removing that rule. It looks like neither WooCommerce nor WooCommerce Admin have it (they have plenty of `@todo` comments around the code), so I guess we should copy them.

Anyway, I still think if we can handle to-do tasks with GitHub issues, it's better than in-line comments.

nerrad replied:
See https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/984

nerrad commented on assets/js/settings/shared/get-setting.js:
```diff
@@ -0,0 +1,29 @@
+ * @export
+ * @param {string} name                               The identifier for the
+ *                                                    setting.
+ * @param {mixed}  [fallback=false]                   The value to use as a
```
hmm I'll check that.  Linting didn't fail for me but if this is a true spacing problem (not just a github display issue) then ya I'll fix.

nerrad commented on assets/js/settings/shared/get-setting.js:
```diff
@@ -0,0 +1,29 @@
+ * @param {mixed}  [fallback=false]                   The value to use as a
+ *                                                    fallback if the setting is
+ *                                                    not in the state.
+ * @param {function} [validation=( value ) => value]  A callback to use for
```
Ya I like that suggestion.  I'll make the change and fix spacing.

nerrad commented on assets/js/settings/shared/test/set-setting.js:
```diff
@@ -0,0 +1,24 @@
+	it( 'should replace existing value', () => {
+		const original = getSetting( 'adminUrl' );
+		setSetting( 'adminUrl', 'not original' );
+		expect( getSetting( 'adminUrl' ) ).not.toBe( original );
```
In this case, we are ensuring that the fallback is _not_ being used (because the actual value is present).  So the current test is correct.

nerrad commented on src/Assets.php:
```diff
@@ -16,19 +16,21 @@ class Assets {
 	 * Initialize class features on init.
+	 *
+	 * @since $VID:$ Moved most initialization to BootStrap and AssetDataRegistry
+	 *               classes as a part of ongoing refactor
```
Yup it will.  Not sure how we should handle that (improving the script will be tricky).  Maybe just catch lint issues in subsequent pulls after the automatic replacement?

nerrad replied:
Alternatively, for scenarios like this, we could start the comment immediately below the tag?  I'm not sure if that will cause linting errors on it's own but would solve the alignment issue.

Aljullu replied:
> Alternatively, for scenarios like this, we could start the comment immediately below the tag?
That sounds like the best solution to me.

nerrad replied:
See e573975

nerrad commented:
@Aljullu this is ready for another review.  I think I've addressed all feedback.  Post merge, I'll review all the comments and make sure there are issues for each of the follow-ups.

Aljullu commented on src/Assets/Api.php:
```diff
@@ -0,0 +1,146 @@
+	public function __construct( Package $package ) {
+		$this->package = $package;
+	}
+	/**
```
Nitpicking here, but I would add a blank line here...
```suggestion

	/**
```

Aljullu commented on src/Assets/Api.php:
```diff
@@ -0,0 +1,146 @@
+		return $this->package->get_url( $relative_path );
+	}
+
+
```
...and remove it from here.
```suggestion
```

Aljullu commented on src/Assets/Api.php:
```diff
@@ -0,0 +1,146 @@
+	}
+
+
+
```
Probably we can remove a couple of blank lines here?

Aljullu commented on src/Assets/Api.php:
```diff
@@ -0,0 +1,146 @@
+	 * @param array  $extra_dependencies   Extra dependencies to be explicitly
+	 *                                     added to the generated array.
+	 *
+	 * @return array  An arry of dependencies
```
Typo:
```suggestion
	 * @return array  An array of dependencies
```

Aljullu commented on src/Assets/AssetDataRegistry.php:
```diff
@@ -0,0 +1,269 @@
+	 */
+	private $data = [];
+
+
```
Double blank lines again. :slightly_smiling_face: 
```suggestion
```

Aljullu commented on src/Package.php:
```diff
@@ -7,45 +7,36 @@
-
-	/**
-	 * Stores if init has ran yet.
+	 * For back compat this is provided. Ideally, you should use register your
```
There seems to be an unnecessary word here:
```suggestion
	 * For back compat this is provided. Ideally, you should register your
```

mikejolley commented on src/Domain/Bootstrap.php:
```diff
@@ -0,0 +1,151 @@
+	}
+
+	/**
+	 * Remove core blocks (for 3.6 and below).
```
Core uses packages now, so only older versions of woo core need use this code.

nerrad replied:
> Core uses packages

Ahh, my initial thought for this block was that this ensured that when the plugin is active it overrode any blocks in core (which it _did_ for the older blocks that were included in core as opposed to this package).

Which leads to the question - when the plugin is active, how does it override the package included in core?

mikejolley replied:
The package system in woo core ensures the latest version is loaded - that being either the version in core, or the version in the feature plugin.

This code takes care of legacy versions of core (<3.6) where blocks are included regardless of this feature plugin being installed, before packages were implemented.

nerrad replied:
Follow-up to revert this change?

mikejolley replied:
Ill do it and expand the docblock
